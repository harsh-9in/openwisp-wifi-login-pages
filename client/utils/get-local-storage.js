const localStorage = () =>
  window.useLocalStorageMock ? window.localStorageMock : window.localStorage;

export default localStorage();
