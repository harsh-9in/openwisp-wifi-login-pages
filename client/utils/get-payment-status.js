import axios from "axios";
import {t} from "ttag";
import {paymentStatusUrl} from "../constants";
import logError from "./log-error";

const getPaymentStatus = async (orgSlug, paymentId, oneTimeToken) => {
  const url = paymentStatusUrl(orgSlug, paymentId);
  try {
    const response = await axios({
      method: "get",
      headers: {
        "content-type": "application/json",
        Authorization: `status_token ${oneTimeToken}`,
      },
      url,
    });
    if (response.status === 200) {
      return response.data.status;
    } else {
      logError("Cannot get payment status");
      return false;
    }
  } catch (error) {
    logError(error, t`ERR_OCCUR`);
    return false;
  }
};

export default getPaymentStatus;
