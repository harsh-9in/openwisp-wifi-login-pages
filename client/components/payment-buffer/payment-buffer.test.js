/* eslint-disable prefer-promise-reject-errors */
import {shallow} from "enzyme";
import React from "react";
import PropTypes from "prop-types";
import {loadingContextValue} from "../../utils/loading-context";
import PaymentBuffer from "./payment-buffer";
import tick from "../../utils/tick";
import getPaymentStatus from "../../utils/get-payment-status";
import loadTranslation from "../../utils/load-translation";

jest.mock("axios");
jest.mock("../../utils/get-payment-status");
jest.mock("../../utils/load-translation");

const createTestProps = (props) => ({
  orgSlug: "default",
  settings: {oneTimeTokenName: "status_token"},
  location: {
    search:
      "?status_token=MagicToken&payment_id=af906397-64e3-4343-ae4e-3c6ee416a205",
  },
  ...props,
});

describe("Test <PaymentBuffer /> URL query_param absent", () => {
  let props;
  let wrapper;
  const originalLog = console.log;

  beforeAll(() => {
    delete window.location;
    window.location = {
      search: "",
    };
  });

  beforeEach(() => {
    props = createTestProps();
    PaymentBuffer.contextTypes = {
      setLoading: PropTypes.func,
    };
    console.log = jest.fn();
    console.error = jest.fn();
    loadTranslation("en", "default");
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
    console.log = originalLog;
  });

  it("should redirect unauthenticated user to login", async () => {
    props = createTestProps();
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual("/default/login");

    window.location.search = "?status_token=MagicToken";
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual("/default/login");

    window.location.search = "?payment_id=af906397-64e3-4343-ae4e-3c6ee416a205";
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual("/default/login");
  });

  it("should redirect authenticated user to status", async () => {
    props = createTestProps({
      location: {search: "?status_token=MagicToken"},
      isAuthenticated: true,
    });
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual("/default/status");

    window.location.search = "?status_token=MagicToken";
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual("/default/status");

    window.location.search = "?payment_id=af906397-64e3-4343-ae4e-3c6ee416a205";
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual("/default/status");
  });
});

describe("Test <PaymentBuffer /> query_param present", () => {
  let props;
  let wrapper;
  const originalLog = console.log;

  beforeAll(() => {
    delete window.location;
    window.location = {
      search:
        "?status_token=MagicToken&payment_id=af906397-64e3-4343-ae4e-3c6ee416a205",
    };
  });

  beforeEach(() => {
    props = createTestProps();
    PaymentBuffer.contextTypes = {
      setLoading: PropTypes.func,
    };
    getPaymentStatus.mockClear();
    console.log = jest.fn();
    console.error = jest.fn();
    loadTranslation("en", "default");
  });

  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    jest.restoreAllMocks();
    console.log = originalLog;
  });

  it("should redirect authenticated user payment status", async () => {
    props = createTestProps({isAuthenticated: true});
    getPaymentStatus.mockReturnValue("success");
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    await tick();
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual(
      "/default/payment/success",
    );

    getPaymentStatus.mockReturnValue("failed");
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    await tick();
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual(
      "/default/payment/failed",
    );

    getPaymentStatus.mockReturnValue("waiting");
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    await tick();
    expect(wrapper.find("Redirect").length).toEqual(1);
    expect(wrapper.find("Redirect").props().to).toEqual(
      "/default/payment/draft",
    );
  });

  it("should show payment status to unauthenticated user", async () => {
    props = createTestProps();
    getPaymentStatus.mockReturnValue("success");
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    await tick();
    expect(wrapper).toMatchSnapshot();

    getPaymentStatus.mockReturnValue("failed");
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    await tick();
    expect(wrapper).toMatchSnapshot();

    getPaymentStatus.mockReturnValue("waiting");
    wrapper = shallow(<PaymentBuffer {...props} />, {
      context: loadingContextValue,
    });
    await tick();
    expect(wrapper).toMatchSnapshot();
  });
});
