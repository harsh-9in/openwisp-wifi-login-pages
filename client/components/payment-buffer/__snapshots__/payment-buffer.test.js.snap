// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test <PaymentBuffer /> query_param present should show payment status to unauthenticated user 1`] = `
<div
  className="container content"
>
  <div
    className="inner payment-buffer"
  >
    <div
      className="main-column single"
    >
      <div
        className="inner"
      >
        <h2
          className="row"
        >
           
          Your transaction was successful.
           
        </h2>
        <div
          className="row"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Kindly login to start using wifi service.",
            }
          }
        />
        <div
          className="row"
        >
          <Link
            className="button full"
            to="/default/login/"
          >
            Sign In
          </Link>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test <PaymentBuffer /> query_param present should show payment status to unauthenticated user 2`] = `
<div
  className="container content"
>
  <div
    className="inner payment-buffer"
  >
    <div
      className="main-column single"
    >
      <div
        className="inner"
      >
        <h2
          className="row"
        >
           
          Your transaction failed.
           
        </h2>
        <div
          className="row"
          dangerouslySetInnerHTML={
            Object {
              "__html": "You can start a new transaction after signing in.",
            }
          }
        />
        <div
          className="row"
        >
          <Link
            className="button full"
            to="/default/login/"
          >
            Sign In
          </Link>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test <PaymentBuffer /> query_param present should show payment status to unauthenticated user 3`] = `
<div
  className="container content"
>
  <div
    className="inner payment-buffer"
  >
    <div
      className="main-column single"
    >
      <div
        className="inner"
      >
        <h2
          className="row"
        >
           
          You transaction is pending.
           
        </h2>
        <div
          className="row"
          dangerouslySetInnerHTML={
            Object {
              "__html": "Sign-in to complete the transaction.",
            }
          }
        />
        <div
          className="row"
        >
          <Link
            className="button full"
            to="/default/login/"
          >
            Sign In
          </Link>
        </div>
      </div>
    </div>
  </div>
</div>
`;
